// testcontainers/dist/src/docker-compose/docker-compose-client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDockerComposeClient = void 0;
const logger_1 = require("../logger");
const docker_compose_1 = __importStar(require("docker-compose"));
const docker_compose_options_1 = require("./docker-compose-options");
function getDockerComposeClient(compat) {
    switch (compat) {
        case undefined:
            return new MissingDockerComposeClient();
        case "v1":
            return new DockerComposeV1Client();
        case "v2":
            return new DockerComposeV2Client();
    }
}
exports.getDockerComposeClient = getDockerComposeClient;
class DockerComposeV1Client {
    up(options, services) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (services) {
                    logger_1.log.info(`Upping DockerCompose environment services ${services.join(", ")}...`);
                    yield docker_compose_1.default.upMany(services, yield (0, docker_compose_options_1.defaultDockerComposeOptions)(options));
                }
                else {
                    logger_1.log.info(`Upping DockerCompose environment...`);
                    yield docker_compose_1.default.upAll(yield (0, docker_compose_options_1.defaultDockerComposeOptions)(options));
                }
                logger_1.log.info(`Upped DockerCompose environment`);
            }
            catch (err) {
                yield handleAndRethrow(err, (error) => __awaiter(this, void 0, void 0, function* () {
                    try {
                        logger_1.log.error(`Failed to up DockerCompose environment: ${error.message}`);
                        yield this.down(options, { removeVolumes: true, timeout: 0 });
                    }
                    catch (_a) {
                        logger_1.log.error(`Failed to down DockerCompose environment after failed up`);
                    }
                }));
            }
        });
    }
    pull(options, services) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (services) {
                    logger_1.log.info(`Pulling DockerCompose environment images "${services.join('", "')}"...`);
                    yield docker_compose_1.default.pullMany(services, yield (0, docker_compose_options_1.defaultDockerComposeOptions)(Object.assign(Object.assign({}, options), { logger: logger_1.pullLog })));
                }
                else {
                    logger_1.log.info(`Pulling DockerCompose environment images...`);
                    yield docker_compose_1.default.pullAll(yield (0, docker_compose_options_1.defaultDockerComposeOptions)(Object.assign(Object.assign({}, options), { logger: logger_1.pullLog })));
                }
                logger_1.log.info(`Pulled DockerCompose environment`);
            }
            catch (err) {
                yield handleAndRethrow(err, (error) => __awaiter(this, void 0, void 0, function* () { return logger_1.log.error(`Failed to pull DockerCompose environment images: ${error.message}`); }));
            }
        });
    }
    stop(options) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                logger_1.log.info(`Stopping DockerCompose environment...`);
                yield docker_compose_1.default.stop(yield (0, docker_compose_options_1.defaultDockerComposeOptions)(options));
                logger_1.log.info(`Stopped DockerCompose environment`);
            }
            catch (err) {
                yield handleAndRethrow(err, (error) => __awaiter(this, void 0, void 0, function* () { return logger_1.log.error(`Failed to stop DockerCompose environment: ${error.message}`); }));
            }
        });
    }
    down(options, downOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                logger_1.log.info(`Downing DockerCompose environment...`);
                yield docker_compose_1.default.down(Object.assign(Object.assign({}, (yield (0, docker_compose_options_1.defaultDockerComposeOptions)(options))), { commandOptions: dockerComposeDownCommandOptions(downOptions) }));
                logger_1.log.info(`Downed DockerCompose environment`);
            }
            catch (err) {
                yield handleAndRethrow(err, (error) => __awaiter(this, void 0, void 0, function* () { return logger_1.log.error(`Failed to down DockerCompose environment: ${error.message}`); }));
            }
        });
    }
}
class DockerComposeV2Client {
    up(options, services) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (services) {
                    logger_1.log.info(`Upping DockerCompose environment services ${services.join(", ")}...`);
                    yield docker_compose_1.v2.upMany(services, yield (0, docker_compose_options_1.defaultDockerComposeOptions)(options));
                }
                else {
                    logger_1.log.info(`Upping DockerCompose environment...`);
                    yield docker_compose_1.v2.upAll(yield (0, docker_compose_options_1.defaultDockerComposeOptions)(options));
                }
                logger_1.log.info(`Upped DockerCompose environment`);
            }
            catch (err) {
                yield handleAndRethrow(err, (error) => __awaiter(this, void 0, void 0, function* () {
                    try {
                        logger_1.log.error(`Failed to up DockerCompose environment: ${error.message}`);
                        yield this.down(options, { removeVolumes: true, timeout: 0 });
                    }
                    catch (_a) {
                        logger_1.log.error(`Failed to down DockerCompose environment after failed up`);
                    }
                }));
            }
        });
    }
    pull(options, services) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (services) {
                    logger_1.log.info(`Pulling DockerCompose environment images "${services.join('", "')}"...`);
                    yield docker_compose_1.v2.pullMany(services, yield (0, docker_compose_options_1.defaultDockerComposeOptions)(Object.assign(Object.assign({}, options), { logger: logger_1.pullLog })));
                }
                else {
                    logger_1.log.info(`Pulling DockerCompose environment images...`);
                    yield docker_compose_1.v2.pullAll(yield (0, docker_compose_options_1.defaultDockerComposeOptions)(Object.assign(Object.assign({}, options), { logger: logger_1.pullLog })));
                }
                logger_1.log.info(`Pulled DockerCompose environment`);
            }
            catch (err) {
                yield handleAndRethrow(err, (error) => __awaiter(this, void 0, void 0, function* () { return logger_1.log.error(`Failed to pull DockerCompose environment images: ${error.message}`); }));
            }
        });
    }
    stop(options) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                logger_1.log.info(`Stopping DockerCompose environment...`);
                yield docker_compose_1.v2.stop(yield (0, docker_compose_options_1.defaultDockerComposeOptions)(options));
                logger_1.log.info(`Stopped DockerCompose environment`);
            }
            catch (err) {
                yield handleAndRethrow(err, (error) => __awaiter(this, void 0, void 0, function* () { return logger_1.log.error(`Failed to stop DockerCompose environment: ${error.message}`); }));
            }
        });
    }
    down(options, downOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                logger_1.log.info(`Downing DockerCompose environment...`);
                yield docker_compose_1.v2.down(Object.assign(Object.assign({}, (yield (0, docker_compose_options_1.defaultDockerComposeOptions)(options))), { commandOptions: dockerComposeDownCommandOptions(downOptions) }));
                logger_1.log.info(`Downed DockerCompose environment`);
            }
            catch (err) {
                yield handleAndRethrow(err, (error) => __awaiter(this, void 0, void 0, function* () { return logger_1.log.error(`Failed to down DockerCompose environment: ${error.message}`); }));
            }
        });
    }
}
class MissingDockerComposeClient {
    up() {
        throw new Error("DockerCompose is not installed");
    }
    pull() {
        throw new Error("DockerCompose is not installed");
    }
    stop() {
        throw new Error("DockerCompose is not installed");
    }
    down() {
        throw new Error("DockerCompose is not installed");
    }
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function handleAndRethrow(err, handle) {
    return __awaiter(this, void 0, void 0, function* () {
        const error = err instanceof Error ? err : new Error(err.err.trim());
        yield handle(error);
        throw error;
    });
}
function dockerComposeDownCommandOptions(options) {
    const result = [];
    if (options.removeVolumes) {
        result.push("-v");
    }
    if (options.timeout) {
        result.push("-t", `${options.timeout / 1000}`);
    }
    return result;
}
