// testcontainers/dist/src/modules/hivemq/hivemq-container.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StartedHiveMQContainer = exports.HiveMQContainer = void 0;
const path = __importStar(require("path"));
const generic_container_1 = require("../../generic-container/generic-container");
const abstract_started_container_1 = require("../abstract-started-container");
const wait_1 = require("../../wait-strategy/wait");
const START_LOG_MESSAGE_REGEX = /(.*)Started HiveMQ in(.*)/i;
const HIVEMQ_BASE_PATH = "/opt/hivemq";
const MQTT_PORT = 1883;
class HiveMQContainer extends generic_container_1.GenericContainer {
    constructor(image = "hivemq/hivemq-ce:2023.5") {
        super(image);
        this.temporaryFileSystemOptions = new Map();
        this.withExposedPorts(MQTT_PORT)
            .withWaitStrategy(wait_1.Wait.forLogMessage(START_LOG_MESSAGE_REGEX))
            .withStartupTimeout(120000);
        this.addFileToFileSystemOptions("log");
        this.addFileToFileSystemOptions("data");
        this.withTmpFs(Object.fromEntries(this.temporaryFileSystemOptions));
    }
    addFileToFileSystemOptions(filename) {
        this.temporaryFileSystemOptions.set(path.join(HIVEMQ_BASE_PATH, filename), "rw");
    }
    start() {
        const _super = Object.create(null, {
            start: { get: () => super.start }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return new StartedHiveMQContainer(yield _super.start.call(this));
        });
    }
}
exports.HiveMQContainer = HiveMQContainer;
class StartedHiveMQContainer extends abstract_started_container_1.AbstractStartedContainer {
    constructor(startedTestContainer) {
        super(startedTestContainer);
        this.port = startedTestContainer.getMappedPort(MQTT_PORT);
    }
    getPort() {
        return this.port;
    }
    getConnectionString() {
        return `mqtt://${this.getHost()}:${this.getPort()}`;
    }
}
exports.StartedHiveMQContainer = StartedHiveMQContainer;
